test/dynamo/test_repros.py

```diff
@@ -3595,6 +3595,51 @@ def fn(x=None):
         self.assertEqual(cnt.frame_count, 4)
 
 
+    def test_tensor_set_data(self):
+        # https://github.com/pytorch/pytorch/issues/113030
+        def func1(x, y):
+            x.data = y
+            x.add_(1)
+            return x
+
+        def func2(x, y):
+            x.data = y
+            y.data = torch.zeros([0])
+            return x
+
+        def func3(x, y):
+            z = x
+            x.data = y
+            y.data = torch.zeros([0])
+            return x is z
+
+        for backend in ["eager", "aot_eager", "inductor"]:
+            torch._dynamo.reset()
+            for func in [func1, func2, func3]:
+                if func is func1 and backend != "eager":
+                    # NYI - aot_autograd with add_() on data
+                    return
+                cnt = torch._dynamo.testing.CompileCounterWithBackend(backend)
+
+                compiled_fn = torch.compile(func, backend=cnt)
+                for i in range(0, 5):
+                    # Inputs
+                    eager_a = torch.rand([6])
+                    compiled_a = torch.clone(eager_a)
+                    eager_b = torch.rand([6])
+                    compiled_b = torch.clone(eager_b)
+
+                    # Eager
+                    func(eager_a, eager_b)
+                    # Compiled
+                    compiled_fn(compiled_a, compiled_b)
```

Comment:
self.assertEqual(func(..), compiled_fn(..))

For func3 return

Reply:
yea, adding some bwd changes too